### PR TITLE
Unify separate atomic counter functions in runtime

### PR DESCRIPTION
**Pull request by NickBarnes**
This small PR unifies two separate implementations of atomic counters in the runtime.
There are functions for `atomic_uintnat` counters in `atomic_refcount.h`, used in `bigarray.c`, and we use similar counters in `major_gc.c`, with one inline function `atomic_fetch_add_verify_ge0` in `platform.h` which is only used for decrement but has an assertion to guard against underflow. I unify the two into inline functions in `camlatomic.h` (which seems like the natural home for them).
Incidentally, the increment/decrement functions now return the value _after_ the update—obviously more intuitive than the previous `atomic_refcount` semantics—which means a `1` in `bigarray.c` has become a `0`.
There's also a tiny tweak in `misc.h` to avoid the circular header dependency which otherwise arises.

### CODE REVIEW

File: runtime/caml/camlatomic.h [hunk 1]
```diff
@@ -56,6 +57,31 @@ typedef _Atomic intnat atomic_intnat;
...
+  atomic_store_release(counter, n);
+}
+
+/* Decrement an atomic uintnat counter. Assertion check for
+ * underflow. Returns the new value. */
+
+Caml_inline uintnat caml_atomic_counter_decr(atomic_uintnat* counter)
+{
+  uintnat old = atomic_fetch_sub(counter, 1);
+  CAMLassert ((intnat)old > 0);
```
Comment (gasche) [inline runtime/caml/camlatomic.h]: Noob question, sorry: why do we cast from `uintnat` to `intnat` before performing this comparison? (And also `old+1` below.)

**Review by gasche — APPROVED** [review]
This looks fine to me -- see minor question inline.

My impression is that this PR leaves the code slightly inconsistent in that some of the `num_domains_to_*` variables use some counter operations and also some plain loads and stores. There are some calls to `caml_atomic_store_release(&num_domain_to_*` that could be turned into `caml_atomic_refcount_init` calls, and it may also make sense to define a helper function to encapsulate the `caml_atomic_load_acquire(&foo) == 0` logic, which is frequent in major_gc.c, and where the acquire ordering might be important for performance.

**Review by MisterDA — APPROVED** [review]

File: runtime/caml/camlatomic.h (hunk 1 above)
Comment (MisterDA) [reply runtime/caml/camlatomic.h]: Well we cannot test if an unsigned number is negative… but we could test for `old != CAML_UINTNAT_MAX`, which means the counter would have looped (counter was 0).

Comment (gasche) [reply runtime/caml/camlatomic.h]: My understanding is that we are testing that the pre-decrement value was *strictly positive* (not zero), which also makes sense on unsigned integers.

Comment (NickBarnes) [reply runtime/caml/camlatomic.h]: Good catch; I originally wrote this test differently, as a test on the new value (which, as @MisterDA notes, can never be negative) and the cast survived a tweak. I'm removing both casts.

**Comment by gasche** [comment]
(I had some questions/discussions in my comment and my expectation would be that maybe @NickBarnes has an opinion on them, which I would be curious to hear -- maybe he decides to update the PR as a result -- before we decide to merge.)

**Comment by NickBarnes** [comment]
> There are some calls to `caml_atomic_store_release(&num_domain_to_*` that could be turned into `caml_atomic_refcount_init` calls

Updating these.

> and it may also make sense to define a helper function to encapsulate the `caml_atomic_load_acquire(&foo) == 0` logic, which is frequent in major_gc.c, and where the acquire ordering might be important for performance.

As I understand it, `atomic_load(p)` is the same as `atomic_load_explicit(p, memory_order_seq_cst)` which is the same as `atomic_load_explicit(p, memory_order_acquire)` which is the same as `atomic_load_acquire(p)`. I assume the reason we have `atomic_load_acquire()` at all is to explicitly document the memory order for the programmer. I'm replacing all accesses to these atomic counters with calls to the appropriate inline `caml_atomic_counter_` functions.

> My impression is that this PR leaves the code slightly inconsistent in that some of the `num_domains_to_*` variables use some counter operations and also some plain loads and stores.

Fixing these up for consistency. Thank you.

**Comment by NickBarnes** [comment]
Updated to reflect review comments: all access to atomic counters is now through `caml_atomic_counter_...` inline functions.
As an experiment, I also made [this branch](https://github.com/NickBarnes/ocaml/tree/nick-atomic-counter-check) which uses the singleton-struct hack to force all users to use this `camlatomic.h` API. That found an instance in `otherlibs/unix/mmap_ba.c` which was just using `--`, which I have here replaced with the appropriate function call. Atomic types in C are defined such that `--` and `++` decrement and increment operators on them do the right thing; this whole PR is based on the assumption that we avoid using those to emphasize the atomic nature - and cost - of the operations to the programmer.

**Pull request merged**